pkg/webhook/servicecatalog/serviceinstance/validation/handler.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"context"
+	"net/http"
+
+	sc "github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1beta1"
+	"github.com/kubernetes-incubator/service-catalog/pkg/webhookutil"
+
+	admissionTypes "k8s.io/api/admission/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type Validator func(ctx context.Context, req admission.Request, si *sc.ServiceInstance, l *webhookutil.TracedLogger) error
+
+// AdmissionHandler handles ServiceInstance
+type AdmissionHandler struct {
+	decoder *admission.Decoder
+	client  client.Client
+
+	CreateValidators []Validator
+	UpdateValidators []Validator
+}
+
+var _ admission.Handler = &AdmissionHandler{}
+var _ admission.DecoderInjector = &AdmissionHandler{}
+var _ inject.Client = &AdmissionHandler{}
+
+func NewAdmissionHandler() {
+	h := &AdmissionHandler{}
+	h.UpdateValidators = []Validator{h.DenyPlanChangeIfNotUpdatable}
+}
+
+// Handle handles admission requests.
+func (h *AdmissionHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
```

Comment:
I have also another idea to simplify that (but add also magic)

```go
// AdmissionHandler handles ServiceInstance
type AdmissionHandler struct {
	Validations map[admissionTypes.Operation][]Validator
}
```

```go
validators, found := h.Validations[admissionTypes.Create]
if !found {
	traced.Infof("ServiceInstance AdmissionHandler wehbook does not support action %q", req.Operation)
	return admission.Allowed("action not taken")
}
for _, fn := range validators {
	if err := fn(ctx, req, si, traced); err != nil {
		errs = append(errs, err)
	}
}
```

@adamwalach @piotrmiskiewicz  what do you thing about that? this really simply the the Handler it self. It also working if you initialize the `&AdmissionHandler{}` without creating a map because this `validators, found := h.Validations[admissionTypes.Create]` return also not found if `Validations` is nil

Reply:
When I'm looking at such ideas I'm more and more convinced to have completely separate admission handlers.

Reply:
separate for creating and updating actions?